internal/cephfs/clone.go

```diff
@@ -0,0 +1,188 @@
+	if cloneErr != nil {
+		klog.Errorf(util.Log(ctx, "failed to clone snapshot %s %s to %s %v"), volID, snapshotID, cloneID, cloneErr)
+		return cloneErr
+	}
+	var clone CloneStatus
+	clone, cloneErr = getcloneInfo(ctx, volOpt, cr, cloneID)
+	if cloneErr != nil {
+		return cloneErr
+	}
+	if clone.Status.State == cephFSCloneInprogress {
+		return ErrCloneInProgress{err: fmt.Errorf("clone is in progress for %v", cloneID)}
+	}
+	if clone.Status.State == cephFSCloneFailed {
+		cloneErr = fmt.Errorf("clone %s is in %s state", cloneID, clone.Status.State)
+	}
+	if clone.Status.State == cephFSCloneComplete {
+		// This is a work around to fix sizing issue for cloned images
+		err = resizeVolume(ctx, volOpt, cr, cloneID, volOpt.Size)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to expand volume %s: %v"), cloneID, err)
+			return err
+		}
+		// As we completed clone, remove the intermediate snap
+		if err = unprotectSnapshot(ctx, parentvolOpt, cr, snapshotID, volID); err != nil {
+			// Incase the snap is already unprotected we get ErrSnapProtectionExist error code
+			// in that case we are safe and we could discard this error and we are good to go
+			// ahead with deletion
+			if !errors.Is(err, ErrSnapProtectionExist) {
+				klog.Errorf(util.Log(ctx, "failed to unprotect snapshot %s %v"), snapshotID, err)
+				return err
+			}
+		}
+		if err = deleteSnapshot(ctx, parentvolOpt, cr, snapshotID, volID); err != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapshotID, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func cleanupCloneFromSubvolumeSnapshot(ctx context.Context, volID, cloneID volumeID, parentVolOpt *volumeOptions, cr *util.Credentials) error {
+	// snapshot name is same as clone name as we need a name which can be
+	// identified during PVC-PVC cloning.
+	snapShotID := cloneID
+	snapInfo, err := getSnapshotInfo(ctx, parentVolOpt, cr, snapShotID, volID)
+	if err != nil {
+		if errors.Is(err, ErrSnapNotFound) {
+			return nil
+		}
+		return err
+	}
+
+	if snapInfo.Protected == "true" {
+		err = unprotectSnapshot(ctx, parentVolOpt, cr, snapShotID, volID)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to unprotect snapshot %s %v"), snapShotID, err)
+			return err
+		}
+	}
+	err = deleteSnapshot(ctx, parentVolOpt, cr, snapShotID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapShotID, err)
+		return err
+	}
+	return nil
+}
+
+func createCloneFromSnapshot(ctx context.Context, parentVolOpt, volOptions *volumeOptions, vID *volumeIdentifier, sID *snapshotIdentifier, cr *util.Credentials) error {
+	snapID := volumeID(sID.FsSnapshotName)
+	err := cloneSnapshot(ctx, parentVolOpt, cr, volumeID(sID.FsSubvolName), snapID, volumeID(vID.FsSubvolName), volOptions)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			var ecip ErrCloneInProgress
+			if !errors.As(err, &ecip) {
```

Comment:
need to use `errors.Is()` here too

Reply:
same here, its already taken care and the reivew is on an outdated content. :)